include/unifex/async_auto_reset_event.hpp

```diff
@@ -0,0 +1,150 @@
+// (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.
+
+#pragma once
+
+#include <unifex/async_manual_reset_event.hpp>
+#include <unifex/defer.hpp>
+#include <unifex/just_done.hpp>
+#include <unifex/just_void_or_done.hpp>
+#include <unifex/let_value.hpp>
+#include <unifex/let_value_with.hpp>
+#include <unifex/let_value_with_stop_token.hpp>
+#include <unifex/sender_concepts.hpp>
+
+#include <mutex>
+
+namespace unifex {
+
+namespace _aare {
+/**
+ * A *Stream* backed by an auto-reset event.
+ */
+struct async_auto_reset_event final {
+  /**
+   * Initializes an async_auto_reset_event in the unset state.
+   */
+  async_auto_reset_event() noexcept : async_auto_reset_event(false) {}
+
+  /**
+   * Initializes an async_auto_reset_event.
+   *
+   * If startReady is true, the event is initialized in the set state; otherwise
+   * it's initialized in the unset state.
+   */
+  explicit async_auto_reset_event(bool startReady) noexcept
+    : state_(startReady ? state::SET : state::UNSET)
+    , event_(startReady) {}
+
+  /**
+   * Puts the event in the set state if it's not in the done state.
+   *
+   * Signals any waiting *Senders* if the event was unset.
+   */
+  void set() noexcept;
+
+  /**
+   * Puts the event in the done state.
+   *
+   * Signals any waiting *Senders* if the event was unset.
+   */
+  void set_done() noexcept;
+
+  struct stream_view;
+  /**
+   * Retrieves a *Stream*-shaped view of the event.
+   */
+  stream_view stream() noexcept;
+
+private:
+  enum class state { UNSET, SET, DONE };
+  /**
+   * Alternatives to using `mutex`:
+   *   1. "2 stage commit" with `atomic<state>`, requires re-try
+   *   2. drop `set_done()` support, undesirable
+   * @see discussion on D34156868 for reasoning.
```

Comment:
This isn't super useful upstream.  Assuming it can be made public, it would be better to inline the reasoning here.